index.rst

```diff
@@ -42,20 +42,163 @@
+Additionally, there is a "grouping" concept that can be used by scripts to indicate that observations taken with the same group identifier are related.[*]_
+The group ID is how standard visits will be observed with both snaps being part of the same group.
+The group ID will be provided to each script by the ScriptQueue and will be uniquely defined for that execution of the script.
+The system will guarantee that no observing script will ever be given the same group ID.
+
+Since DM software always mediates data access via a Data Butler this leads to a requirement that raw data must always be retrievable from a butler with a Data ID solely containing the obsid.
+Additionally the detector name can be specified in order to process a subset of the acquired data, and if a group ID is given instead of an obsid, multiple observations can be processed together.
+
+All data products generated using this system are treated as transient data products that will not be archived and will not be required to include permanent provenance tracking.
+The purpose is to provide early feedback to observing in the form of data quality assessments.
+
+.. [*] In :cite:`LTS-836` the group ID is called a "transaction ID" since it is associated with a single queue transaction.
+
+.. note::
+
+  No LSST rapid analysis pipelines are required to co-add data incrementally.
+  Pipelines should not co-add data as it arrives and should always defer co-addition until the final observation in the group has been taken.
+
+Controlling Data Processing from Observing Scripts
+==================================================
+
+In the initial design of the OCS Controlled Pipeline System (OCPS) [*]_ the pipeline processing will be controlled directly by commands sent from the observing script to a CSC.
+The commands will indicate which data to process and the name of the pipeline to execute but they will not include algorithmic code.
+The data will be specified using a "dataId" to match standard Data Butler usage.
+Once the pipeline completes an event will be published.
+
+.. [*] In :cite:`LDM-148` this is called the "OCS-Controlled Batch Service."
+
+The sequence of events for a typical observing script would then be:
+
+* Send a command to the camera to take an image.
+  This command should include the relevant groupId string as supplied from the ScriptQueue, and the observing type (such as science, bias, or flat).
+* Wait for the event indicating that the exposure has been written to the OODS (see :cite:`DMTN-111`).
+  This event will include the relevant obsid.
+* Send a command to the OCPS requesting that this obsid be processed.
+  This command should include the name of the pipeline to execute.
+* Acquire more data and send processing requests for each one.
+* When all data are acquired and single-frame processing has completed, send a command to the OCPS requesting that a co-add pipeline be executed with all data using the shared groupId.
```

Comment:
I think calling it "co-add" is confusing.  This is the code that does the gather step, but it needn't coadd.
Also, do we want to require that the per-obsid and the final pipelines be distinct?  Logically we have one pipeline that sits there processing data, and finally performs some other step.  This way guarantees that we have to do I/O for the processes to co-operate (maybe we can avoid this with butler magic?).

Reply:
@TallJimbo had a scheme whereby we could have a super clever system that is told how many exposures will arrive and has the gather step wait on the data and then have it trigger the gather automatically.  I think I had that in an earlier draft for the data-driven processing. Do we want to attempt to do that first? It sounded like the gen3 infrastructure wasn't yet able to wait for additional data to arrive.  It seemed much simpler for the script to explicitly command single frame processing and when it was done with that send the combine command.